File: ios/Classes/OrtMagika.swift
import Flutter
import Foundation
import onnxruntime_objc
import os.log

/// OrtVad is a class responsible for performing voice activity detection (VAD) using ONNX Runtime.
class OrtMagika {
  /// The path to the ONNX model file.
  private let modelPath: String

  /// Lazy instantiation of OrtSessionObjects used to manage the ONNX session.
  private lazy var model: OrtSessionObjects = {
    OrtSessionObjects(modelPath: modelPath, includeOrtExtensions: false)!
  }()

  /// Initializes the OrtVad instance with a provided model path.
  /// - Parameter modelPath: The path to the ONNX model file.
  init(modelPath: String) {
    self.modelPath = modelPath
  }

  /// Performs inference using the ONNX model.
  /// - Parameters:
  ///   - fileBytes: The input data as an array of floats. 1536 ranging from 0 to 255 (bytes)
  ///     and 256 (representing padding).
  /// - Returns: A list of Float32 values representing the probability of each class.
  func doInference(fileBytes: [Float]) -> FlutterStandardTypedData? {
    do {
      let session = model.session
      let bytesTensor = try createORTValue(
        from: fileBytes, elementType: .float,
        shape: [NSNumber(value: 1), NSNumber(value: fileBytes.count)])
      let inputs = [
        "bytes": bytesTensor
      ]
      let outputName = "target_label"
      let outputNames = Set([outputName])

      let outputs = try session.run(withInputs: inputs, outputNames: outputNames, runOptions: nil)
      guard let rawOutputValue = outputs[outputName] else {
        os_log("Output value not found")
        return nil
      }
      let rawOutputData = try rawOutputValue.tensorData() as Data
      // Model output is 32-bit floats (known from model inspection, see Netron)
      // See here for info on Swift type choice:
      // https://docs.flutter.dev/platform-integration/platform-channels?tab=type-mappings-swift-tab
      os_log("Output data: %{public}s", rawOutputData.description)
      let outputFloats = FlutterStandardTypedData(float32: rawOutputData)
      return outputFloats
    } catch {
      os_log("Error in doInference: %{public}s", error.localizedDescription)
      for symbol in Thread.callStackSymbols {
        os_log("Stack trace: %{public}s", symbol)
      }
      return nil
    }
  }
}

File: android/src/main/kotlin/com/telosnex/fonnx/OrtMagika.kt
package com.telosnex.fonnx

import ai.onnxruntime.*
import android.util.Log

class OrtMagika(private val modelPath: String) {
    private val model: OrtSessionObjects by lazy {
        OrtSessionObjects(modelPath, false)
    }

    fun doInference(fileBytes: FloatArray): FloatArray? {
        try {
            val ortEnv = model.ortEnv
            val session = model.ortSession
            val batchSize = 1
            val inputs = mutableMapOf<String, OnnxTensor>()
            inputs["bytes"] = OnnxTensor.createTensor(ortEnv, arrayOf(fileBytes))
            val outputName = "target_label"
            val outputNames = setOf(outputName)
            session?.let {
                val startTime = System.currentTimeMillis()
                val rawResult = it.run(inputs, outputNames)
                val endTime = System.currentTimeMillis()
                Log.d("[OrtMagika.kt]", "Inference time: ${endTime - startTime} ms")
                val targetLabel = rawResult.get(outputName).get().value as? Array<FloatArray>
                return targetLabel?.get(0)
            }
            return null
        } catch (e: Exception) {
            Log.e("[OrtMagika.kt]", "Error in doInference: ${e.message}")
            return null
        }
    }
}
